posts/2023-09-30-keycloak-with-openliberty.adoc

```diff
@@ -0,0 +1,126 @@
+
+Now, with this annotation set up, your REST endpoints can be protected using the `@RolesAllowed` annotation which will trigger the authentication flow when a user tries to access the endpoint.
+After authentication, the user's access token can be obtained using the `OpenIdContext`.
+
+This example shows a JAX-RS resource which contains a `/username` endpoint which is only accessible by users with the `admin` role and a `/os` endpoint which is accessible by both users with the `admin` role and users with the `user` role.
+
+[source, java]
+----
+@ApplicationScoped
+@Path("/system/properties")
+public class SystemResource {
+
+    @Inject
+    @RestClient
+    private SystemService systemService;
+
+    @Inject
+    private OpenIdContext openIdContext;
+
+    @GET
+    @Path("/username")
+    @RolesAllowed({ "admin" })
+    public String getUsername() {
+        return systemService.getUsername(openIdContext.getAccessToken().getToken());
+    }
+
+    @GET
+    @Path("/os")
+    @RolesAllowed({ "admin", "user" })
+    public String getOS() {
+        return systemService.getOS(openIdContext.getAccessToken().getToken());
+    }
+    
+}
+----
+
+Once the requests to these endpoints are authenticated and authorized, the endpoint can now use the access token of the authenticated user.
+In the above example, the access token is used as a bearer token to make a request to another protected resource by including it in the request header in the format `Authorization: Bearer <access-token>`.
+The next section will demonstrate how this bearer token can be consumed by an Open Liberty application using MicroProfile JWT to protect its resources.
+
+== Consuming an access token from Keycloak using MicroProfile JWT
+
+link:https://openliberty.io/docs/latest/reference/feature/mpJwt-2.1.html[MicroProfile JWT] can easily be used to consume access tokens sent as bearer tokens.
+
+This example shows the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#jwt[MicroProfile Config properties] required to validate an access token issued by the `openliberty` realm in Keycloak.
+
+[source, text]
+----
+mp.jwt.verify.issuer=http://localhost:8080/realms/openliberty
+mp.jwt.verify.publickey.location=http://localhost:8080/realms/openliberty/protocol/openid-connect/certs
+----
+
+In this example, the `mp.jwt.verify.issuer` is the endpoint of the `openliberty` realm and the `mp.jwt.verify.publickey.location` is the JSON Web Key Sets (JWKS) endpoint of the `openliberty` realm.
+
+Now, MicroProfile JWT is setup to validate access tokens issued by the `openliberty` realm sent as bearer tokens to resources protected using the `@RolesAllowed` annotation.
+
+Similarly to the previous section, this example shows a JAX-RS resource which contains a `/username` endpoint only accessible by users with the `admin` role and a `/os` endpoint accessible by both users with the `admin` role and users with the `user` role.
```

Comment:
This paragraph is a little confusing in that we're referring to both the section before and the section after, which makes it hard to read/follow. Is there a way that we could maybe better set up this whole section to make this transition easier/make more sense. So maybe introduce both options up front, then go through each one?